geometric_kernels/spaces/special_orthogonal_group.py

```diff
@@ -0,0 +1,408 @@
+"""
+Special orthogonal group, SO(n).
+
+The code is courtеsy of Iskander Azangulov (*) and Andrei Smolensky (*)
+
+(*) St. Petersburg University
+"""
+
+import itertools as it
+import math
+import operator
+from functools import reduce
+from typing import Any, Dict
+
+import geomstats as gs
+import lab as B
+import numpy as np
+from opt_einsum import contract as einsum
+
+from geometric_kernels.eigenfunctions import Eigenfunctions
+from geometric_kernels.lab_extras import (
+    from_numpy,
+    isclose,
+    swapaxes,
+    take_along_last_axis,
+)
+from geometric_kernels.spaces import DiscreteSpectrumSpace
+
+
+def fixed_length_partitions(n, L):  # noqa
+    """
+    https://www.ics.uci.edu/~eppstein/PADS/IntegerPartitions.py
+    Integer partitions of n into L parts, in colex order.
+    The algorithm follows Knuth v4 fasc3 p38 in rough outline;
+    Knuth credits it to Hindenburg, 1779.
+    """
+
+    # guard against special cases
+    if L == 0:
+        if n == 0:
+            yield []
+        return
+    if L == 1:
+        if n > 0:
+            yield [n]
+        return
+    if n < L:
+        return
+
+    partition = [n - L + 1] + (L - 1) * [1]
+    while True:
+        yield partition.copy()
+        if partition[0] - 1 > partition[1]:
+            partition[0] -= 1
+            partition[1] += 1
+            continue
+        j = 2
+        s = partition[0] + partition[1] - 1
+        while j < L and partition[j] >= partition[0] - 1:
+            s += partition[j]
+            j += 1
+        if j >= L:
+            return
+        partition[j] = x = partition[j] + 1
+        j -= 1
+        while j > 0:
+            partition[j] = x
+            s -= x
+            j -= 1
+        partition[0] = s
+
+
+class SOEigenfunctions(Eigenfunctions):
+    """Eigenfunctions for Special Orthogonal Group SO(n)."""
+
+    def __init__(self, dim: int, num_representations: int = 10):
+        r"""
+        :param dim: dimensionality in which the group acts.
+        :param num_representation: number of representations of the group to compute. Values larger than 10 are impractical.
+        """
+        assert (
+            dim != 4
+        ), "Dimension 4 is not supported since SO(4) is not a semisimple group."
```

Comment:
Is this an issue for some applications?

Reply:
Solely sanity checks, for the dimension 4 the math does not work.